src/nvim/diff.c

```diff
@@ -2644,16 +3421,31 @@ void ex_diffgetput(exarg_T *eap)
   dprev = NULL;
 
   for (dp = curtab->tp_first_diff; dp != NULL;) {
+
+    if (dp->is_linematched && !eap->addr_count) {
+      // handle the case with overlapping diff blocks
+      while (dp->is_linematched && dp->df_next && (dp->df_next->df_lnum[idx_cur] ==
+            dp->df_lnum[idx_cur] + dp->df_count[idx_cur]) && dp->df_next->df_lnum[idx_cur] ==
+          (eap->line1 + off + 1)) 
+      {
+        dp = dp->df_next;
+      }
+    }
+
     if (dp->df_lnum[idx_cur] > eap->line2 + off) {
       // past the range that was specified
       break;
     }
     dfree = NULL;
+
     lnum = dp->df_lnum[idx_to];
     count = dp->df_count[idx_to];
 
     if ((dp->df_lnum[idx_cur] + dp->df_count[idx_cur] > eap->line1 + off)
         && (u_save(lnum - 1, lnum + count) != FAIL)) {
+      if (dp->is_linematched && !eap->addr_count) {
+        breakafteronehunk = 1;
+      }
```

Comment:
Does this need a comment?

Reply:
I just removed this. I found that this is actually redundant. The case which this would fix is also fixed by the while loop above the for loop that this is in.
https://github.com/neovim/neovim/blob/9c31f950684d21bed3639e9effe967f46678ef9b/src/nvim/diff.c#L3429-L3438